src/include/miopen/env.hpp

```diff
@@ -26,195 +26,193 @@
+                value_ = value.value();
+            }
         }
-
-        return false; // shouldn't reach here
     }
-};
-
-// Supports hexadecimals (with leading "0x"), octals (if prefix is "0") and decimals (default).
-// Returns 0 if environment variable is in wrong format (strtoull fails to parse the string).
-template <>
-struct ParseEnvVal<uint64_t>
-{
-    static uint64_t go(const char* vp) { return std::strtoull(vp, nullptr, 0); }
-};
-
-template <>
-struct ParseEnvVal<std::string>
-{
-    static std::string go(const char* vp) { return std::string{vp}; }
-};
 
-template <typename T>
-struct EnvVar
-{
-private:
-    T value{};
-    bool is_unset = true;
-
-public:
-    const T& GetValue() const { return value; }
-
-    bool IsUnset() const { return is_unset; }
-
-    void Unset() { is_unset = true; }
-
-    void UpdateValue(const T& val)
+    template <typename U>
+    U value(std::optional<U> alternate_value = std::nullopt) const
+    {
+        return static_cast<U>(
+            value_.value_or(alternate_value.value_or(static_cast<U>(default_value_))));
+    }
+    void clear()
     {
-        is_unset = false;
-        value    = val;
+        clearEnvironmentVariable(name_);
+        value_.reset();
     }
 
-    explicit EnvVar(const char* const name, const T& def_val)
+    bool exist() const { return value_.has_value(); }
+
+    template <typename U>
+    void update(U value)
     {
-        // NOLINTNEXTLINE (concurrency-mt-unsafe)
-        const char* vp = std::getenv(name);
-        if(vp != nullptr) // a value was provided
+        if constexpr(is_same_v<U, std::string> || is_same_v<U, std::string_view>)
         {
-            is_unset = false;
-            value    = ParseEnvVal<T>::go(vp);
+            setEnvironmentVariable(name_, value);
         }
-        else // no value provided, use default value
+        if constexpr(is_same_v<U, bool>)
         {
-            value = def_val;
+            setEnvironmentVariable(name_, value ? "1" : "0");
         }
+        if constexpr(std::is_integral_v<remove_cvref_t<U>> && !is_same_v<U, bool>)
+        {
+            setEnvironmentVariable(name_, std::to_string(value));
+        }
+        value_ = value;
     }
+    std::string_view name() const { return name_; }
+
+private:
+    std::string_view name_;
+    T default_value_;
+    std::optional<T> value_{std::nullopt};
 };
 
-} // end namespace internal
+} // end namespace detail
 
-// static inside function hides the variable and provides
-// thread-safety/locking
-// Used in global namespace
-#define MIOPEN_DECLARE_ENV_VAR(name, type, default_val)                            \
-    namespace miopen::env {                                                        \
-    struct name                                                                    \
+#define MIOPEN_DECLARE_ENV_VAR(_name, _type, ...)                                  \
+    [[maybe_unused]] static const struct __struct_##_name                          \
```

Comment:
That sort of declaration is bloating binary size, because it has a unique copy of that struct in each file where the same env var is used.
Just a
```
struct _name
{
};
```
combines the same struct declaration into one copy in the binary.
That is a critical for the tests where we use `MIOPEN_TEST_ALL` env variable almost in every test and then link everything into one binary.

Also it impacts performance since it has a unique copy of the singleton the same env variable and all the copies will be initialized individually. Seems that's why you've introduced `create_if_missing` flag to synchronize all the copies.

Reply:
+1 each envvar object should be defined only once.

Reply:
> +1 each envvar object should be defined only once.

We've already discussed this.